jax/experimental/jax2tf/examples/tflite/mnist/mnist.py

```diff
@@ -0,0 +1,134 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from absl import app
+from absl import flags
+
+from jax.experimental import jax2tf
+from jax.experimental.jax2tf.examples import mnist_lib
+
+import numpy as np
+
+import tensorflow as tf
+import tensorflow_datasets as tfds
+
+flags.DEFINE_string('tflite_file_path', '/tmp/jax2tf/mnist/mnist.tflite',
+                    'Path where to save the TensorFlow Lite file.')
+flags.DEFINE_integer('serving_batch_size', 1,
+                     ('For what batch size to prepare the serving signature. '))
+flags.DEFINE_integer('num_epochs', 10, 'For how many epochs to train.')
+
+FLAGS = flags.FLAGS
+
+
+# A helper function to evaluate the TF Lite model using "test" dataset.
+def evaluate_tflite_model(tflite_model, test_ds):
+  # Initialize TFLite interpreter using the model.
+  interpreter = tf.lite.Interpreter(model_content=tflite_model)
+  interpreter.allocate_tensors()
+  input_tensor_index = interpreter.get_input_details()[0]['index']
+  output = interpreter.tensor(interpreter.get_output_details()[0]['index'])
+
+  # Run predictions on every image in the "test" dataset.
+  prediction_digits = []
+  labels = []
+  for image, one_hot_label in test_ds:
+    interpreter.set_tensor(input_tensor_index, image)
+
+    # Run inference.
+    interpreter.invoke()
+
+    # Post-processing: remove batch dimension and find the digit with highest
+    # probability.
+    digit = np.argmax(output()[0])
+    prediction_digits.append(digit)
+    labels.append(np.argmax(one_hot_label))
+
+  # Compare prediction results with ground truth labels to calculate accuracy.
+  accurate_count = 0
+  for index in range(len(prediction_digits)):
+    if prediction_digits[index] == labels[index]:
+      accurate_count += 1
+  accuracy = accurate_count * 1.0 / len(prediction_digits)
+  return accuracy
+
+
+def main(_):
+  logging.info('Loading the MNIST TensorFlow dataset')
+  train_ds = mnist_lib.load_mnist(
+      tfds.Split.TRAIN, batch_size=mnist_lib.train_batch_size)
+  test_ds = mnist_lib.load_mnist(
+      tfds.Split.TEST, batch_size=FLAGS.serving_batch_size)
+
+  (flax_predict,
+   flax_params) = mnist_lib.FlaxMNIST.train(train_ds, test_ds, FLAGS.num_epochs)
+
+  def predict(image):
+    return flax_predict(flax_params, image)
+
+  # Convert Flax model to TF function.
+  tf_predict = tf.function(
+      jax2tf.convert(predict, with_gradient=False, enable_xla=False),
+      input_signature=[
+          tf.TensorSpec(
+              shape=[FLAGS.serving_batch_size, 28, 28, 1],
```

Comment:
I am curious to what extent is the conversion sensitive to the batch size. Can it work for a batch size greater than 1 and in that case must the converted code be invoked with an appropriately-sized batch?

Reply:
Good question. TFLite interpreter now supports dynamic batch size. We can convert the model with shape=[None, 28, 28, 1], and then dynamic resize_input_tensor inside the TFLite interpreter. However, for jax2tf to work, we need to use batch shape polymorphism. Maybe should just reuse the saved_model_lib and convert to tflite from the saved model, so that we have shape polymorphism automatically enabled.